tutorials/tour/_posts/2017-02-13-annotations.md

```diff
@@ -11,79 +11,51 @@ next-page: default-parameter-values
+    if (x == 1) accumulator else factorialHelper(x - 1, accumulator * x)
+  }
+  factorialHelper(x, 1)
 }
 ```
-
-The following Java program prints out the contents of the file whose name is passed as the first argument to the `main` method.
-
+The `factorialHelper` method has the `@tailrec` which ensures the method is indeed tail-recursive. If we were to change the implementation of `factorialHelper` to the following, it would fail:
 ```
-package test;
-import examples.Reader;  // Scala class !!
-public class AnnotaTest {
-    public static void main(String[] args) {
-        try {
-            Reader in = new Reader(args[0]);
-            int c;
-            while ((c = in.read()) != -1) {
-                System.out.print((char) c);
-            }
-        } catch (java.io.IOException e) {
-            System.out.println(e.getMessage());
-        }
-    }
+def factorial(x: Int): Int = {
+  @tailrec
+  def factorialHelper(x: Int): Int = {
+    if (x == 1) 1 else x * factorialHelper(x - 1)
+  }
+  factorialHelper(x)
 }
 ```
+We would get the message "Recursive call not in tail position".
 
-Commenting out the `throws` annotation in the class Reader produces the following error message when compiling the Java main program:
 
-```
-Main.java:11: exception java.io.IOException is never thrown in body of
-corresponding try statement
-        } catch (java.io.IOException e) {
-          ^
-1 error
-```
+## Annotations affecting code generation
+Some annotations like `@inline` affect the generated code. (e.g. your jar file might have different bytes than if you hadn't used the annotation). Inlining code means that the compiler finds the calls to a function definition and moves the body of the function to all the places where the method is called. This can increase performance but also increase the size of the generated code. Using the annotation `@inline` does not ensure that a method will be inlined, but it will cause the compiler to do it iff some heuristics about the size of the generated code are met.
 
 ### Java Annotations ###
+This section is for Java developers.
+**Note:** Make sure you use the `-target:jvm-1.8` option with Java annotations.
 
-**Note:** Make sure you use the `-target:jvm-1.5` option with Java annotations.
-
-Java 1.5 introduced user-defined metadata in the form of [annotations](http://java.sun.com/j2se/1.5.0/docs/guide/language/annotations.html). A key feature of annotations is that they rely on specifying name-value pairs to initialize their elements. For instance, if we need an annotation to track the source of some class we might define it as
+Java 1.5 introduced user-defined metadata in the form of [annotations](https://docs.oracle.com/javase/tutorial/java/annotations/). A key feature of annotations is that they rely on specifying name-value pairs to initialize their elements. For instance, if we need an annotation to track the source of some class we might define it as
```

Comment:
I'd suggest replacing "Java 1.5 introduced" with "Java has" or "Java also has" — the introduction is now ancient history.